### PR TITLE
styled-components: Use CommonJS export for ESM + module: Node16

### PR DESCRIPTION
**Pull request by karlhorky**
Fix `@types/styled-components` for ESM (eg. `"type": "module"` in `package.json`) and modern Node.js module resolution (`"module": "Node16"` or `"module": "NodeNext"` in `tsconfig.json`)

Playground demos:

- [Playground before this PR](https://www.typescriptlang.org/play?target=99&moduleResolution=99&module=100#code/JYWwDg9gTgLgBAcgAIwJ5gKYGcD0W0A2GAJgLQDGE4EAdhjTFggFCiSxz6pHFwBmUKoi48KVSHQZNmzSjXxwAymACGNOAF5OhEgDosqmgAMjQA)
- [Playground after this PR](https://www.typescriptlang.org/play?&dtPR=65150&install-plugin=playground-dt-review#code/JYWwDg9gTgLgBAcgAIwJ5gKYGcD0W0A2GAJgLQDGE4EAdhjTFggFCiSxz6pHFwBmUKoi48KVSHQZNmzSjXxwAymACGNOAF5OhEgDosqmgAMjQA)

Before this PR:

Screenshot 2023-04-16 at 18 59 24

Error message:

```
Property 'span' does not exist on type 'typeof import("file:///node_modules/@types/styled-components/index")'.
```

cc @andrewbranch

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: no URL, this fixes broken types
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by karlhorky** [comment]
Ok I wrapped this in a namespace now, simplified diff without whitespace changes here:

- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65150/files?w=1

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## styled-components ([<kbd>unpkg</kbd>](https://unpkg.com/browse/styled-components@latest/))
was missing the following properties:
1. StyleSheetConsumer
1. StyleSheetContext
1. version
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 061ea2bf2332a2f1f6462002327a56cf45ae156e
</p>

**Comment by karlhorky** [comment]
Oh no 😆 Tests are passing but the types are still broken for `Node16` / `NodeNext` - please don't merge yet!

**Comment by karlhorky** [comment]
Well I've tried a few things now, but I guess I am not understanding how to write CommonJS exports for these type declarations.

Should be something like this:

```js
import styled, { // no error
   css,  // no error
   span, // error: Module 'styled-components' has no exported member 'span'. Did you mean to use 'import span from "styled-components"' instead?
} from 'styled-components';

styled.span`` // no error
styled.css // error: Property 'css' does not exist on type 'StyledInterface'
css // no error
```

- `css`, etc. can be imported as a named import, but not as a property on the default import styled
- `span` (and all other HTML elements) can be accessed as a property on `styled` but not as a named import

Not sure how to do this with CommonJS-style `export =` syntax. The things I've tried so far are:

- switch `export default` to `export =`
- wrap everything in `declare namespace` and then `export =` for that namespace
- wrap in `declare module styled`

File: types/styled-components/index.d.ts [hunk 1]
```diff
@@ -20,421 +20,423 @@ declare global {
     interface ShadowRoot {}
 }
 
-import * as CSS from 'csstype';
-import * as React from 'react';
-import * as hoistNonReactStatics from 'hoist-non-react-statics';
+declare module 'styled-components' {
```
Comment (Andarist) [inline types/styled-components/index.d.ts:23]: what do u want to solve with this? unless there is some strong reason to wrap the content like this - i would avoid this syntax

**Comment by karlhorky** [comment]
Maybe is not so easy!

> yes, `styled-components` could "proxy" to its CJS with import condition using an .mjs wrapper~. That's how we want to fix the very same issue for Emotion. Then you could adjust the types package like i suggested as that would match the new runtime behavior of SC in node

- https://twitter.com/AndaristRake/status/1648000936688730113

**Comment by karlhorky** [comment]
Ok, thanks for the additional hint @andrewbranch! Closing this until any changes in `styled-components` itself appear.

**Pull request closed without merging**